File: engine/core/Decimation.ts
/**
 * SCUMM's actor scaling, in the engine. Errata ruling 24.
 *
 * `smallCostumeScaleTable` is an eight-bit reversal -- 0, 128, 64, 192, 32,
 * ... -- walked one entry per source row and column. A line is drawn when its
 * entry is under the scale and skipped when it is over. NOTHING IS BLENDED,
 * which is the whole reason this stays crisp where a resample smears: every
 * drawn row is still exactly one source row.
 *
 * The distribution is what makes it work. Consecutive entries are as far
 * apart in value as possible, so the lines dropped at any scale are spread
 * evenly through the figure instead of clumping. It is an ordered dither
 * applied to a sequence rather than a plane.
 *
 * This file is the exact counterpart of tools/pixelart/decimation.py. If the
 * two ever disagree the sprite in the room stops matching the sprite that was
 * measured, and the measured threshold stops meaning anything.
 */

const TABLE_SIZE = 256;

function bitReversalTable(): number[] {
  const table: number[] = [];
  for (let n = 0; n < TABLE_SIZE; n += 1) {
    let bits = 0;
    for (let shift = 0; shift < 8; shift += 1) {
      if (n & (1 << shift)) bits |= 1 << (7 - shift);
    }
    table.push(bits);
  }
  return table;
}

export const SCALE_TABLE = bitReversalTable();

/** Which of `count` source lines survive at `scale` (0-255). */
export function kept(count: number, scale: number): number[] {
  const lines: number[] = [];
  for (let index = 0; index < count; index += 1) {
    if ((SCALE_TABLE[index % TABLE_SIZE] as number) < scale) lines.push(index);
  }
  return lines;
}

/**
 * The scale whose decimation is closest to `wanted` lines.
 *
 * Searched, not computed. The table is not linear over short runs -- forty
 * rows sample only its first forty entries, whose distribution is coarse --
 * so the scale that yields 26 rows out of 40 is not 26/40 of 255.
 */
export function scaleFor(count: number, wanted: number): number {
  let best = 255;
  let bestGap = Number.POSITIVE_INFINITY;
  for (let scale = 1; scale < TABLE_SIZE; scale += 1) {
    const gap = Math.abs(kept(count, scale).length - wanted);
    if (gap < bestGap) {
      best = scale;
      bestGap = gap;
    }
    if (bestGap === 0 && gap > 0) break;
  }
  return best;
}

/**
 * Decimates a region of `source` into a new canvas of the kept lines.
 *
 * Pixels are copied through ImageData rather than drawn with drawImage, so
 * there is no filtering step to disable and no chance of a browser deciding
 * to be helpful. Rows and columns walk the table independently, which is what
 * keeps a figure's proportions from shearing.
 */
export function decimate(
  source: CanvasImageSource,
  sx: number,
  sy: number,
  width: number,
  height: number,
  scale: number,
): HTMLCanvasElement {
  const rows = kept(height, scale);
  const columns = kept(width, scale);
  const cut = document.createElement('canvas');
  cut.width = Math.max(1, columns.length);
  cut.height = Math.max(1, rows.length);
  const target = cut.getContext('2d');
  if (!target) return cut;
  target.imageSmoothingEnabled = false;

  const whole = document.createElement('canvas');
  whole.width = width;
  whole.height = height;
  const from = whole.getContext('2d');
  if (!from) return cut;
  from.imageSmoothingEnabled = false;
  from.drawImage(source, sx, sy, width, height, 0, 0, width, height);

  const read = from.getImageData(0, 0, width, height);
  const write = target.createImageData(cut.width, cut.height);
  for (let y = 0; y < rows.length; y += 1) {
    for (let x = 0; x < columns.length; x += 1) {
      const at = (((rows[y] as number) * width) + (columns[x] as number)) * 4;
      const to = ((y * cut.width) + x) * 4;
      write.data[to] = read.data[at] as number;
      write.data[to + 1] = read.data[at + 1] as number;
      write.data[to + 2] = read.data[at + 2] as number;
      write.data[to + 3] = read.data[at + 3] as number;
    }
  }
  target.putImageData(write, 0, 0);
  return cut;
}

File: tools/pixelart/decimation.py
"""SCUMM's actor scaling, tested against ours. Evidence for ruling 15.

WHAT SCUMM ACTUALLY DID. ScummVM's source scales actors CONTINUOUSLY and stays
crisp by DECIMATING rather than resampling: `smallCostumeScaleTable` is a
256-entry table and the engine walks it one entry per source row and column,
drawing that row or column when the entry is under the scale and skipping it
when it is over. Nothing is blended. A 40px actor at scale 166 is 26 rows of
the original 40, chosen by the table, each still exactly one source row.

The table is a BIT-REVERSAL distribution -- 0, 128, 64, 192, 32, 160, ... --
which is what makes it work. Consecutive entries are as far apart in value as
possible, so the rows it drops at any scale are spread evenly through the
figure rather than clumping. It is the same idea as an ordered dither, applied
to a sequence instead of a plane.

WHAT RULING 15 ASSUMED. Ruling 15 fixed three drawn heights and forbade
interpolation, and the stated reason was smearing. That reason was about
RESAMPLING, which is not what SCUMM did. This module exists to find out
whether the objection survives when the mechanism is right.

THE TEST. Decimate the 40px Thad to 26 and 32 with a bit-reversal table, put
each beside the hand-corrected version, and look at the face -- the earlier
finding was that reducing by ratio lands on the eyes, and a distribution table
does not know where the eyes are either.

Evidence only. Nothing here changes what the game draws.
"""

from __future__ import annotations

import actor
from canvas import IndexedCanvas
from palette import Palette


def bit_reversal_table(size: int = 256) -> list[int]:
    """SCUMM's smallCostumeScaleTable, regenerated rather than transcribed.

    Eight-bit reversal: entry n is n's bits in the opposite order. The result
    is 0, 128, 64, 192, 32, 160, 96, 224 ... -- every prefix of it spread as
    evenly over 0..255 as a prefix of that length can be.
    """
    table = []
    for n in range(size):
        bits = 0
        for shift in range(8):
            if n & (1 << shift):
                bits |= 1 << (7 - shift)
        table.append(bits)
    return table


TABLE = bit_reversal_table()


def kept(count: int, scale: int) -> list[int]:
    """Which of `count` source lines survive at `scale` (0-255).

    One table entry per source line, in order, exactly as the interpreter
    walks it. A line is drawn when its entry is under the scale.
    """
    return [index for index in range(count) if TABLE[index % len(TABLE)] < scale]


def decimate(source: IndexedCanvas, scale: int) -> IndexedCanvas:
    """Scale by dropping whole rows and columns. No blending anywhere.

    Rows and columns use the same table but are walked independently, which
    is what SCUMM did and what keeps a figure's proportions from shearing.
    """
    rows = kept(source.height, scale)
    columns = kept(source.width, scale)
    out = IndexedCanvas(max(1, len(columns)), max(1, len(rows)), fill=actor.TRANSPARENT)
    for target_y, source_y in enumerate(rows):
        for target_x, source_x in enumerate(columns):
            out.put(target_x, target_y, source.pixels[source_y][source_x])
    return out


def scale_for(source_height: int, wanted: int) -> int:
    """The scale value whose decimation is closest to `wanted` rows.

    Searched rather than computed. The table is not linear over short runs --
    40 rows sample only the first 40 entries, whose distribution is coarse --
    so the scale that yields 26 rows out of 40 is not simply 26/40 of 255.
    """
    best, best_gap = 255, None
    for scale in range(1, 256):
        gap = abs(len(kept(source_height, scale)) - wanted)
        if best_gap is None or gap < best_gap:
            best, best_gap = scale, gap
        if best_gap == 0 and gap > 0:
            break
    return best
